### PR TITLE
stretch_cluster.py: Added forced recovery and healthy

### PR DESCRIPTION
**Pull request by pdhiran**
1. Added forced recovery and healthy modes with stretch mode
2. Added selections for mds and client( rgw ) for  Monitor configuration database tests
3. Made changes to include verification steps for bugs:
      a. https://bugzilla.redhat.com/show_bug.cgi?id=2025800
      b. https://bugzilla.redhat.com/show_bug.cgi?id=1952763

Signed-off-by: Pawan Dhiran <pdhiran@redhat.com>

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [x] Create a test case in Polarion reviewed and approved.
- [x] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [x] Review the automation design
- [x] Implement the test script and perform test runs
- [x] Submit PR for code review and approve
- [x] Update Polarin Test with Automation script details and update automation fields
- [x] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by pdhiran** [comment]
Logs : http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-3Z3Q59/

Stretch mode test ran for 2.5 hours and failed, as cluster did not reach active + clean state. ( PG's stuck in peering state )

Bug filed for the issue : https://bugzilla.redhat.com/show_bug.cgi?id=2025800

**Comment by pdhiran** [comment]
Pass log for ` suites/pacific/rados/tier-2_rados_basic_regression.yaml `
: http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-I5840Q/

**Comment by pdhiran** [comment]
Pass log for suites/pacific/rados/tier-2_rados_basic_regression.yaml : http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-I5840Q/

File: ceph/rados/core_workflows.py [hunk 1]
```diff
@@ -868,27 +868,29 @@ def change_osd_state(self, action: str, target: int) -> bool:
...
-                log.info(
-                    f"Performing {action} on osd-{target} on host {onode.hostname}. Command {cmd}"
-                )
-                onode.exec_command(sudo=True, cmd=cmd)
-                # Sleeping for 5 seconds for status to be updated
-                time.sleep(5)
-                return True
-        if not flag:
-            log.error(f"osd-{target} not found on cluster")
+        all_daemons = self.run_ceph_command(cmd="ceph orch ps")
```
Comment (pdhiran) [inline ceph/rados/core_workflows.py:875]: Earlier implementation of ` change_osd_state ` made use of crush items/maps to locate the OSD host. which worked in most of the scenarios. ( made use of command ` sudo ceph osd ls-tree host ` to list all osd's under the host )

But with deployment of stretched clusters, the crush items are modified and the OSD's are moved to new crush items.
Hence made changes to fetch the osd daemon host details from ` ceph orch ps ` command, as this does not change with crush changes on cluster.

Comment (pdhiran) [reply ceph/rados/core_workflows.py:875]: Observed this issue with new suite : http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-RT28OQ/

**Comment by pdhiran** [comment]
Pass log for suite ` suites/pacific/rados/tier-2_rados_stretch_mode.yaml ` : http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-X3ZCUX/

**Comment by pdhiran** [comment]
@psathyan , @SrinivasaBharath , @tintumathew10 , can we have this reviewed? Thanks. I can set up a meet if required.

File: tests/rados/stretch_cluster.py [hunk 2]
```diff
@@ -59,6 +59,65 @@ def run(ceph_cluster, **kw):
...
+            return 1
+
+        log.debug("Starting the stopped OSD's")
+        for osd_id in stop_osds:
+            if not rados_obj.change_osd_state(action="restart", target=osd_id):
+                log.error(f"Unable to restart the OSD : {osd_id}")
+                return 1
+
+        # Sleep for 5 seconds for OSD's to join the cluster
+        time.sleep(5)
```
Comment (psathyan) [inline tests/rados/stretch_cluster.py:104]: We are already having a blind sleep in `change_osd_state`. Do we still want another one here?

File: tests/rados/stretch_cluster.py [hunk 3]
```diff
@@ -510,3 +598,46 @@ def wait_for_alert(node: CephAdmin, alert: str, duration: int):
...
+            False -> fail ( Recovery command could not be executed )
+    """
+    cmd = "ceph osd force_recovery_stretch_mode --yes-i-really-mean-it"
+    try:
+        node.shell([cmd])
+    except Exception:
+        log.error("Recovery with stretch mode could not be triggered")
+        return False
+    log.info("Triggered the recovery in stretch mode")
+    return True
```
Comment (psathyan) [inline tests/rados/stretch_cluster.py:620]: Suggestion

I believe we can extend either `run_ceph_command` or `shell` to handle exceptions gracefully by passing an argument. This way we can avoid doing the above steps below...

**Review by psathyan — COMMENTED** [review]
Could you change the way we are restarting the OSD service?

**Review by tintumathew10 — APPROVED** [review]
LGTM

File: ceph/rados/core_workflows.py [hunk 4]
```diff
@@ -870,25 +870,27 @@ def change_osd_state(self, action: str, target: int) -> bool:
...
+            o_node = [
+                entry["hostname"]
+                for entry in all_daemons
+                if entry["daemon_type"] == "osd" and entry["daemon_id"] == str(target)
+            ][0]
+            host = [node for node in osd_nodes if re.search(o_node, node.hostname)][0]
+        except Exception:
+            log.error(f"Could not find host node for OSD {target}")
             return False
+        cmd = f"systemctl {action} ceph-{cluster_fsid}@osd.{target}.service"
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/core_workflows.py]: Created a new method to fetch the node details for any of the daemon provided for future use.

File: tests/rados/stretch_cluster.py (hunk 3 above)
Comment (pdhiran) [reply tests/rados/stretch_cluster.py]: Handled the same with ` run_ceph_command ` method

**Comment by pdhiran** [comment]
Made all the changes, Will add the pass logs soon. pls review.

File: ceph/rados/core_workflows.py [hunk 5]
```diff
@@ -870,24 +876,46 @@ def change_osd_state(self, action: str, target: int) -> bool:
...
+        log.info(
+            f"Performing {action} on osd-{target} on host {host.hostname}. Command {cmd}"
+        )
+        host.exec_command(sudo=True, cmd=cmd)
+        # Sleeping for 5 seconds for status to be updated
+        time.sleep(5)
+        return True
+
+    def fetch_host_node(self, daemon_type: str, daemon_id: str):
```
Comment (psathyan) [inline ceph/rados/core_workflows.py:893]: Can we make it generic... for example `fetch_host_nodes` and have the argument `daemon_id` as optional?

File: ceph/rados/core_workflows.py [hunk 6]
```diff
@@ -870,24 +876,46 @@ def change_osd_state(self, action: str, target: int) -> bool:
...
+        Args:
+            daemon_type: type of daemon
+                Allowed values: alertmanager, crash, mds, mgr, mon, osd, rgw, prometheus, grafana, node-exporter
+            daemon_id: name of the daemon, ID in case of OSD's
+
+        Returns: ceph object for the node
+
+        """
+        host_nodes = self.ceph_cluster.get_nodes()
+        all_daemons = self.run_ceph_command(cmd="ceph orch ps")
```
Comment (psathyan) [inline ceph/rados/core_workflows.py:905]: We could change this to `ceph orch ps --daemon_type {daemon_type}`.. 
```
command = f"ceph orch ps --daemon_type {daemon_type}"
if daemon_id:
  command += " --daemon_id {daemon_id}"
daemons = self.run_ceph_command(cmd=command)
hosts = [ daemon["hostname"] for daemon in daemons]
return hosts
```

**Review by psathyan — COMMENTED** [review]
I believe we can move the processing to server side

File: ceph/rados/core_workflows.py (hunk 5 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py]: made changes

File: ceph/rados/core_workflows.py (hunk 6 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py]: made changes. moved the search of node to remote cluster

**Comment by pdhiran** [comment]
http://magna002.ceph.redhat.com/ceph-qe-logs/pawan/cephci-run-C3Q7PD/

**Review by psathyan — APPROVED** [review]
LGTM

**Pull request merged**